instrumentation/action_pack/lib/opentelemetry/instrumentation/action_pack/patches/action_controller/metal.rb

```diff
@@ -25,6 +25,10 @@ def dispatch(name, request, response)
               end
 
               super(name, request, response)
+            rescue Exception => e # rubocop:disable Lint/RescueException
```

Comment:
Is this the best place to intercept errors and record errors?

Do the errors that are raised here all translate to HTTP 5xx errors? 

Is it possible that HTTP 4xx responses are also triggered using errors? If so then this should not result in setting the span status to an error. <https://opentelemetry.io/docs/specs/otel/trace/semantic_conventions/http/#status>

Reply:
Hi @arielvalentin , thanks for the feedback.

> Is this the best place to intercept errors and record errors?

If we don't change any code structure or how ActionPack::Instrumentation instruments ActionController, I think add rescue with dispatch function would be a good approach. Otherwise, we can try to prepend different method like [process_action](https://github.com/rails/rails/blob/main/actionpack/lib/action_controller/log_subscriber.rb#L24C9-L24C23). Or just create subscriber to subscribe the log event like some vendor does (e.g. [sentry](https://github.com/getsentry/sentry-ruby/blob/master/sentry-rails/lib/sentry/rails/tracing/action_controller_subscriber.rb#L19-L32))

For current implementation, [dispatch](https://github.com/rails/rails/blob/main/actionpack/lib/action_controller/metal.rb#L224) is calling [process](https://github.com/rails/rails/blob/main/actionpack/lib/action_controller/metal/live.rb#L273), which I think will be the last call sequence of the action controller. 

Since instrumentation prepend dispatch and with super, so it's probably fine to rescue from there. For example, like [show_exceptions.rb](https://github.com/rails/rails/blob/main/actionpack/lib/action_dispatch/middleware/show_exceptions.rb#L30-L41), it also try to rescue from error of @app.call(env).

And like the [comment](https://github.com/open-telemetry/opentelemetry-ruby-contrib/issues/635#issuecomment-1698244337) mentioned, the mode of rails app running will determine if they want to show the exception on web page or raise it internally, but I think it will not affect what dispatch function behave beause the error had already occured. 

For `rescue_from`, then yes, it seems like if user use `rescue_from`, then the error won't be intercepted, but I think user would expect the error, and they can decide what to do with expected error?
<br>

> Do the errors that are raised here all translate to HTTP 5xx errors?

I did some tests against 4xx error, and I think there are two methods to raise 404 with rails

Following methods are direct return 404 without any internal exception/error
```ruby
# 403 forbidden
class UsersController < ApplicationController
  def sample_action
    head: forbidden
    @action.do_action
  end
end
# Same as `head: too_many_requests` and other 4xx error

# 404 by render 404
class UsersController < ApplicationController
  def sample_action
    @action.do_action
    render :status => 404
  end
end

# also get 404 not found by requesting wrong url e.g. 0.0.0.0:8002/whatever_is_wrong_url
```
Above method won't create error span, nor record exception on span.

This method cause 4XX error with direct raise exception
```ruby
def sample_action
  @action.do_action
  raise ActionController::RoutingError.new('Not Found')
end
```

Since it raised the exception, the error will be intercepted, above method will create error span and record exception on span. By looking at this [code](https://github.com/rails/rails/blob/main/actionpack/lib/action_dispatch/routing/route_set.rb#L28-L39), I think when 404 occurs, since it will not get into dispatch function, then it won't make span as error span.

Reply:
This is a potential alternate to this PR
https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/667/files